src/Clip.cpp

```diff
@@ -1053,7 +1116,250 @@ std::shared_ptr<Frame> Clip::apply_effects(std::shared_ptr<Frame> frame)
+
+			// Debug output
+			ZmqLogger::Instance()->AppendDebugMethod("Clip::apply_keyframes (Scale: SCALE_NONE)", "frame->number", frame->number, "source_width", source_size.width(), "source_height", source_size.height());
+			break;
+		}
+	}
+
+	/* GRAVITY LOCATION - Initialize X & Y to the correct values (before applying location curves) */
+	float x = 0.0; // left
+	float y = 0.0; // top
+
+	// Adjust size for scale x and scale y
+	float sx = scale_x.GetValue(frame->number); // percentage X scale
+	float sy = scale_y.GetValue(frame->number); // percentage Y scale
+	float scaled_source_width = source_size.width() * sx;
+	float scaled_source_height = source_size.height() * sy;
+
+	switch (gravity)
+	{
+		case (GRAVITY_TOP_LEFT):
+			// This is only here to prevent unused-enum warnings
+			break;
+		case (GRAVITY_TOP):
+			x = (width - scaled_source_width) / 2.0; // center
+			break;
+		case (GRAVITY_TOP_RIGHT):
+			x = width - scaled_source_width; // right
+			break;
+		case (GRAVITY_LEFT):
+			y = (height - scaled_source_height) / 2.0; // center
+			break;
+		case (GRAVITY_CENTER):
+			x = (width - scaled_source_width) / 2.0; // center
+			y = (height - scaled_source_height) / 2.0; // center
+			break;
+		case (GRAVITY_RIGHT):
+			x = width - scaled_source_width; // right
+			y = (height - scaled_source_height) / 2.0; // center
+			break;
+		case (GRAVITY_BOTTOM_LEFT):
+			y = (height - scaled_source_height); // bottom
+			break;
+		case (GRAVITY_BOTTOM):
+			x = (width - scaled_source_width) / 2.0; // center
+			y = (height - scaled_source_height); // bottom
+			break;
+		case (GRAVITY_BOTTOM_RIGHT):
+			x = width - scaled_source_width; // right
+			y = (height - scaled_source_height); // bottom
+			break;
+	}
+
+	// Debug output
+	ZmqLogger::Instance()->AppendDebugMethod("Clip::apply_keyframes (Gravity)", "frame->number", frame->number, "source_clip->gravity", gravity, "scaled_source_width", scaled_source_width, "scaled_source_height", scaled_source_height);
+
+	/* LOCATION, ROTATION, AND SCALE */
+	float r = rotation.GetValue(frame->number); // rotate in degrees
+	x += (width * location_x.GetValue(frame->number)); // move in percentage of final width
+	y += (height * location_y.GetValue(frame->number)); // move in percentage of final height
+	float shear_x_value = shear_x.GetValue(frame->number);
+	float shear_y_value = shear_y.GetValue(frame->number);
+	float origin_x_value = origin_x.GetValue(frame->number);
+	float origin_y_value = origin_y.GetValue(frame->number);
+
+	bool transformed = false;
+	QTransform transform;
+
+	// Transform source image (if needed)
+	ZmqLogger::Instance()->AppendDebugMethod("Clip::apply_keyframes (Build QTransform - if needed)", "frame->number", frame->number, "x", x, "y", y, "r", r, "sx", sx, "sy", sy);
+
+	if (!isEqual(x, 0) || !isEqual(y, 0)) {
+		// TRANSLATE/MOVE CLIP
+		transform.translate(x, y);
+		transformed = true;
+	}
```

Comment:
The `transformed` boolean isn't really that necessary here. QTransform has an `isIdentity()` method that serves the same purpose. If `transform.isIdentity()` is `false`, some sort of transform was applied. If it's `true`, then the transform hasn't been modified — _**OR**_ all of the modifications canceled each other out, and the _end result_ is a no-op. That's a potential optimization that an external boolean flag can't give you.

(Then again, Qt is most likely efficient enough that if you _do_ pass an identity transform to `QPainter::setTransform()`, it'll just be ignored without affecting either performance or output, so even checking `isIdentity()` is probably unnecessary. I'd be surprised if there's any measurable impact from applying an "unnecessary" identity transform.)

Reply:
Thanks, I've removed the boolean check